src/Memoization.cpp

```diff
@@ -537,22 +537,18 @@ class RewriteMemoizedAllocations : public IRMutator {
 
             // Grab the host pointer argument
             // FIXME: Mutating the buffer_init call in this way is gross.
-            const Call *arg2 = call->args[2].as<Call>();
-            if (arg2 != nullptr && arg2->is_intrinsic(Call::address_of)) {
-                internal_assert(!arg2->args.empty()) << "RewriteMemoizedAllocations: address_of call with zero args.\n";
-                const Load *load = arg2->args[0].as<Load>();
-                if (load != nullptr) {
-                    const IntImm *index = load->index.as<IntImm>();
-
-                    if (index != nullptr && index->value == 0 &&
-                        get_realization_name(load->name) == innermost_realization_name) {
-                        // Everything matches, rewrite _halide_buffer_init to use a nullptr handle for address.
-                        std::vector<Expr> args = call->args;
-                        args[2] = make_zero(Handle());
-                        expr = Call::make(type_of<struct halide_buffer_t *>(), Call::buffer_init,
-                                          args, Call::Extern);
-                        return;
-                    }
+            const AddressOf *arg2 = call->args[2].as<AddressOf>();
+            if (arg2) {
+                const IntImm *index = arg2->index.as<IntImm>();
+
+                if (index != nullptr && index->value == 0 &&
```

Comment:
I know it's in the original code, but this could also just be if (arg2 && is_zero(arg2->index) && get_realization_name...